src/m3ninx/index/segment/fst/encoding/docs/data.go

```diff
@@ -128,3 +128,83 @@ func (r *DataReader) Read(offset uint64) (doc.Document, error) {
 
 	return d, nil
 }
+
+// EncodedDataReader is a reader for the data file for encoded documents.
+type EncodedDataReader struct {
+	data []byte
+}
+
+// NewEncodedDataReader returns a new EncodedDataReader.
+func NewEncodedDataReader(data []byte) *EncodedDataReader {
+	return &EncodedDataReader{
+		data: data,
+	}
+}
+
+// Read reads a doc.EncodedDocument from a data stream starting at the specified offset.
+func (e *EncodedDataReader) Read(offset uint64) (doc.EncodedDocument, error) {
+	if offset >= uint64(len(e.data)) {
+		return doc.EncodedDocument{}, fmt.Errorf(
+			"invalid offset: %v is past the end of the data file", offset,
+		)
+	}
+
+	return doc.EncodedDocument{
+		Bytes: e.data[int(offset):],
+	}, nil
+}
+
+// EncodedDocumentReader is a reader for reading documents from encoded documents.
+type EncodedDocumentReader struct {
+	currFields []doc.Field
+}
+
+// NewEncodedDocumentReader returns a new EncodedDocumentReader.
+func NewEncodedDocumentReader() *EncodedDocumentReader {
+	return &EncodedDocumentReader{}
+}
+
+// Read reads a doc.Document from a doc.EncodedDocument. Returned doc.Document should be
+// processed before calling Read again as the underlying array pointed to by the Fields
+// slice will be updated. This approach avoids allocating a new slice with a new backing
+// array for every document processed, unlike (*DataReader).Read
+func (r *EncodedDocumentReader) Read(encoded doc.EncodedDocument) (doc.Document, error) {
+	r.currFields = r.currFields[:0]
+	id, buf, err := encoding.ReadBytes(encoded.Bytes)
+	if err != nil {
+		return doc.Document{}, err
+	}
+
+	x, buf, err := encoding.ReadUvarint(buf)
+	if err != nil {
+		return doc.Document{}, err
+	}
+	n := int(x)
+
+	var name, val []byte
+	for i := 0; i < n; i++ {
+		name, buf, err = encoding.ReadBytes(buf)
+		if err != nil {
+			return doc.Document{}, err
+		}
+		val, buf, err = encoding.ReadBytes(buf)
+		if err != nil {
+			return doc.Document{}, err
+		}
+		r.currFields = append(r.currFields, doc.Field{
```

Comment:
Would we be able to just overwrite the `r.currFields` on each iteration following the first? That way we don't re-alloc the `doc.Field` struct, and can instead only overwrite the `Value` field (since `Name` should be the same each time). It look like atm we are saving the re-alloc for the array itself, but not the `Field` struct.

Reply:
I think we can actually. We'd need to check if the number of fields we're about to read is greater than the current capacity of r.currFields and allocate a new currFields backing array if so, but I think that's OK as we're relying on implicit expansion right now when that happens while appending.

Reply:
Ahh, so after exploring this, we'd need to do something like this:

```
        ...
	n := int(x)

	if n > cap(r.currFields) {
		r.currFields = make([]doc.Field, n, n*2)
	} else if n <= len(r.currFields) {
		// Clear out any pointers to existing byte slices
		for i := range r.currFields {
			r.currFields[i] = doc.Field{}
		}
		r.currFields = r.currFields[:n]
	} else if n > len(r.currFields) {
		// add n - len(r.currFields) to end of r.currFields
	}

	var name, val []byte
        ...
```

This adds some complexity for a questionable amount of benefit. `append` is just going to copy that `doc.Field` instance anyway so the doc.Field we're allocating here should die at the end of the function call. This leads me to believe it won't be a heap allocation and therefore won't be problematic.

Reply:
Yeah good call. Makes sense then to focus on just avoiding the array allocs